jerry-core/ecma/builtin-objects/ecma-builtin-bigint.c

```diff
@@ -36,6 +53,353 @@
+
+  if (input_bigint_sign && (input_byte_size > bigint_size))
+  {
+    min_size = (input_bits_in_byte > bigint_size) ? input_bits_in_byte : bigint_size;
+  }
+
+  if (min_size < sizeof (uint32_t))
+  {
+    min_size = sizeof (uint32_t);
+  }
+
+  ecma_extended_primitive_t * result_p = ecma_bigint_create ((uint32_t) min_size);
+
+  if (JERRY_UNLIKELY (result_p == NULL))
+  {
+    ecma_deref_bigint (input_bigint_p);
+    return ECMA_VALUE_ERROR;
+  }
+
+  ecma_bigint_digit_t *first_digit_p = ECMA_BIGINT_GET_DIGITS (input_bigint_p, 0);
+  ecma_bigint_digit_t *last_digit_p = ECMA_BIGINT_GET_DIGITS (input_bigint_p, bigint_size);
+
+  /* Calculate the leading zeros of the input_bigint */
+  ecma_bigint_digit_t zeros;
+
+  do
+  {
+    zeros = ecma_big_uint_count_leading_zero (*first_digit_p);
+    ++first_digit_p;
+  }
+  while (first_digit_p < last_digit_p);
+
+  uint32_t bits_of_bigint = (uint32_t) (bigint_size * sizeof (uint64_t)) - zeros;
+
+  uint32_t exact_size = 1;
+
+  if ((bits_of_bigint > (size_of_divisor_in_bits - 1)) || input_bigint_sign)
+  {
+    exact_size = (input_byte_size < bigint_size) ? (uint32_t) input_byte_size : bigint_size;
+
+    if (input_bigint_sign && (input_byte_size > bigint_size))
+    {
+      exact_size = (input_byte_size > bigint_size) ? (uint32_t) input_byte_size : bigint_size;
+    }
+  }
+  else
+  {
+    exact_size = bigint_size;
+  }
+
+  if (input_bigint_sign)
+  {
+    bits_of_bigint += 1;
+  }
+
+  if (bits_of_bigint > (input_bits - 1) || input_bigint_sign)
+  {
+    ecma_bigint_digit_t *digits_p = ECMA_BIGINT_GET_DIGITS (input_bigint_p, 0);
+
+    ecma_bigint_digit_t *digits_end_p = ECMA_BIGINT_GET_DIGITS (input_bigint_p, exact_size);
+
+    ecma_bigint_digit_t *result_number_p = ECMA_BIGINT_GET_DIGITS (result_p, 0);
+
+    int32_t first_cell = 0;
+
+    uint32_t surplus_bits = (whole_part > 0) ? (uint32_t) (whole_part * size_of_divisor_in_bits)
+    : (uint32_t) size_of_divisor_in_bits;
+
+    uint32_t mask_bit = (whole_part == 0) ? (uint32_t) input_bits : (uint32_t) input_bits - surplus_bits;
+
+    if (mask_bit == 0)
+    {
+      mask_bit = size_of_divisor_in_bits - 1;
+    }
+
+    uint32_t check_sign_mask = (uint32_t) 1 << (mask_bit - 1);
+
+    uint32_t mask = (uint32_t) 1 << mask_bit;
+
+    if (mask == 0)
+    {
+      mask = ~((ecma_bigint_digit_t) 0);
+    }
+    else
+    {
+      mask -= 1;
+    }
+
+    uint32_t last_cell = (exact_size >= sizeof (uint32_t)) ? (uint32_t) (min_size / sizeof (uint32_t)) - 1 : 0;
+
+    uint8_t positive_num = 0;
+
+    if (is_signed)
+    {
+      if (input_bigint_sign && ((~digits_p[last_cell] + 1) & check_sign_mask) == 0)
+      {
+        positive_num = 1;
+      }
+
+      uint8_t num_representation = 0;
+
+      if ((digits_p[last_cell] & check_sign_mask) == 0)
+      {
+        num_representation = 1;
+      }
+
+      do
+      {
+        *result_number_p++ = num_representation ? *digits_p++ : ~(*digits_p++);
```

Comment:
So the algorithm does a proper conversion, then turns the result back to negative if needed? Can it be done with a single pass?